Hakawai/Mentions/HKWMentionsStartDetectionStateMachine.m

```diff
@@ -202,22 +191,36 @@ - (void)characterTyped:(unichar)c asInsertedCharacter:(BOOL)inserted previousCha
     }
 }
 
-- (void)deleteTypedCharacter:(unichar)deletedChar withCharacterNowPrecedingCursor:(unichar)precedingChar {
-    // Determine the type of the character
-    enum CharacterType {
-        CharacterTypeSeparator = 0,
-        CharacterTypeNormal
-    };
-    enum CharacterType deletedCharacterType = ([HKWMentionsStartDetectionStateMachine.separatorSet characterIsMember:deletedChar] || deletedChar == 0)
-    ? CharacterTypeSeparator
-    : CharacterTypeNormal;
-    enum CharacterType currentCharacterType = ([HKWMentionsStartDetectionStateMachine.separatorSet characterIsMember:precedingChar] || precedingChar == 0)
-    ? CharacterTypeSeparator
-    : CharacterTypeNormal;
+- (void)deleteTypedCharacter:(unichar)deletedChar
+withCharacterNowPrecedingCursor:(unichar)precedingChar
+                    location:(NSUInteger)location
+                textViewText:(nonnull NSString *)textViewText {
+    // Determine the character types
+    enum CharacterType deletedCharacterType = [self characterTypeOfCharacter:deletedChar];
+    enum CharacterType currentCharacterType = [self characterTypeOfCharacter:precedingChar];
+
+    __strong __auto_type delegate = self.delegate;
 
     switch (self.state) {
         case HKWMentionsStartDetectionStateQuiescentReady: {
-            if (currentCharacterType == CharacterTypeNormal) {
+            BOOL shouldCreateMention = NO;
+            if (location > 1 && textViewText.length > location - 2) {
+                const unichar characterBeforePrecedingChar = [textViewText characterAtIndex:location-2];
+                shouldCreateMention = [HKWMentionsStartDetectionStateMachine.separatorSet characterIsMember:characterBeforePrecedingChar];
```

Comment:
wouldn't this also be true if the mentions character is the first character in the screen? eg, if precedingChar is the first char?

Reply:
also this should more accurately be called "isCharacterBeforePrecedingCharASeparator" or something like that, because we don't know if current character type is a control char or not

Reply:
also also what if the characterbeforeprecedingchar is a control char? what happens then?

Reply:
> wouldn't this also be true if the mentions character is the first character in the screen? eg, if precedingChar is the first char?

Before checking for `characterBeforePrecedingChar`, we are checking for location greater than 1. This would not include the case when preceding character is the first character.
 
We have made changes to set `canCreateMention` (previously `shouldCreateMention`) to true when preceding character is the first character and is a control character.

Reply:
> also this should more accurately be called "isCharacterBeforePrecedingCharASeparator" or something like that, because we don't know if current character type is a control char or not

We have now renamed `shouldCreateMention` as `canCreateMention` as we have now included the check of preceding character type and it tells us if it is possible to create a mention.

Reply:
> also also what if the characterbeforeprecedingchar is a control char? what happens then?

Right now, we are checking for separators (separator set that includes control characters) for character before preceding character and mentions will trigger when this character is a separator(regardless of it being a control character). Please let us know if this behavior is not intended.

Reply:
so a control character is a separator also? is that within the spec of this change? I thought this was just a fix to allow @ mentions after things like parentheses?

Talk to me on slack :)

Reply:
As control characters can overlap with separators, we are treating control characters as separators wherever they do not fulfill conditions of creating a mention.